tensorbay/opendataset/BDD100K/loader.py

```diff
@@ -0,0 +1,120 @@
+
+"""This file defines the BDD100K dataloader."""
+
+import json
+import os
+from typing import Any, Dict
+
+from ...dataset import Data, Dataset
+from ...label import Classification, LabeledBox2D
+
+DATASET_NAME = "BDD100K"
+_IMAGE_FOLDER = "Images"
+_LABELS_FOLDER = "Labels"
+_UPLOAD_FOLDER = "100k"
+MODES = ["test", "train", "val"]
+
+
+def BDD100K(path: str) -> Dataset:
+    """Load BDD100K Dataset to Tensorbay.
+
+    The file structure should be like::
+
+        <path>
+            Images/
+                100k/
+                    test
+                    train
+                    val
+                10k/
+                    test
+                    train
+                    val
+            Labels/
+                bdd100k_labels_images_train.json
+                bdd100k_labels_images_val.json
+
+    Arguments:
+        path: The root directory of the dataset.
+
+    Returns:
+        Loaded :class:`~tensorbay.dataset.dataset.Dataset` instance.
+
+    """
+    root_path = os.path.abspath(os.path.expanduser(path))
+    dataset = Dataset(DATASET_NAME)
+    dataset.load_catalog(os.path.join(os.path.dirname(__file__), "catalog.json"))
+
+    label_path = {
+        "train": os.path.join(root_path, _LABELS_FOLDER, "bdd100k_labels_images_train.json"),
+        "val": os.path.join(root_path, _LABELS_FOLDER, "bdd100k_labels_images_val.json"),
+        "test": "",
+    }
+
+    # add segment
+    for mode in MODES:
+        _load_segment(
+            dataset=dataset,
+            path=root_path,
+            mode=mode,
+            label_file_path=label_path[mode],
+        )
+
+    return dataset
+
+
+def _load_segment(
+    dataset: Dataset,
+    path: str,
+    mode: str,
+    label_file_path: str,
+) -> None:
+    segment = dataset.create_segment(mode)
+    segment_path = os.path.join(path, _IMAGE_FOLDER, _UPLOAD_FOLDER, mode)
+    all_image_name = set(os.listdir(segment_path))  # all names of a segment
+    labeled_image_name = set()  # record the pictures labeled
+
+    if mode == "test":
+        for image_name in all_image_name:
+            data = Data(os.path.join(segment_path, image_name))
+            segment.append(data)
+        return
+
+    with open(label_file_path) as fp:
+        label = json.load(fp)
+
+    for label_dict in label:  # process each picture of a segment
+        image_name = label_dict["name"]
+        labeled_image_name.add(image_name)
+        data = Data(os.path.join(segment_path, image_name))
+        _add_label(data, label_dict)
+        segment.append(data)
+
+    if all_image_name - labeled_image_name:
+        segment = dataset.create_segment("unlabeled_images_for_" + mode)
+
+    for unlabeled_image_name in all_image_name - labeled_image_name:  # process unlabeled images
+        data = Data(os.path.join(segment_path, unlabeled_image_name))
+        segment.append(data)
+
+
+def _add_label(data: Data, label_dict: Dict[str, Any]) -> None:  # add label to each picture
```

Comment:
it will be better to write constant function, which means the function will not modify the input argument, use return value to get the result.
in this case, it is bad to:
```python
_add_label(data, label_dict)  # add label in to label
```
it is better to:
```python
data.label.box2d = _get_label(label_dict) # return the label and do not change the input argument
```